include/ginkgo/core/log/logger.hpp

```diff
@@ -658,10 +664,41 @@ class EnableLogging : public PolymorphicBase {
 
     void clear_loggers() override { loggers_.clear(); }
 
+private:
+    template <size_type Event, typename ConcreteLoggableT, typename = void>
+    struct propagate_log_helper {
+        template <typename... Args>
+        static void propagate_log(const ConcreteLoggableT*, Args&&...)
+        {}
+    };
+
+    template <size_type Event, typename ConcreteLoggableT>
+    struct propagate_log_helper<
+        Event, ConcreteLoggableT,
+        xstd::void_t<decltype(
+            std::declval<ConcreteLoggableT>().get_executor())>> {
+        template <typename... Args>
+        static void propagate_log(const ConcreteLoggableT* loggable,
+                                  Args&&... args)
+        {
+            const auto exec = loggable->get_executor();
+            if (exec->should_propagate_log()) {
+                for (auto& logger : exec->get_loggers()) {
+                    if (logger->needs_propagation()) {
+                        logger->template on<Event>(std::forward<Args>(args)...);
```

Comment:
the propagation is only one direction, right?
when object logs, object will also call those loggers from the corresponding executor (if propagation)

Reply:
yes, there is no propagation in another direction